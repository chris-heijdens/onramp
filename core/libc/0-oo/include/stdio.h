/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDIO_H_INCLUDED
#define __ONRAMP_LIBC_STDIO_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>

#define EOF (-1)

typedef void FILE;

extern FILE* stdin;
extern FILE* stdout;
extern FILE* stderr;

FILE* fopen(const char* filename, const char* mode);
size_t fread(void* ptr, size_t size, size_t nmemb, FILE* file);
size_t fwrite(const void* ptr, size_t size, size_t nmemb, FILE* file);
int fclose(FILE* file);
int feof(FILE* file);

long ftell(FILE* file);
int fseek(FILE* stream, long offset, int base);

#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

int fputc(int c, FILE* file);
int fputs(const char* s, FILE* file);
int putchar(int c);
int puts(const char* s);

/* TODO these are probably not going to be defined by libc/0, only by libc/2 or
 * libc/3
typedef int fpos_t;
int fgetpos(FILE* file, fpos_t* pos);
int fsetpos(FILE* file, const fpos_t* pos);
*/

// TODO only defined by libc/1
int remove(const char* filename);

// posix
int fileno(FILE* file);
typedef int mode_t;
int fchmod(int fileno, int mode); // TODO this belongs in sys/stat.h

#endif
