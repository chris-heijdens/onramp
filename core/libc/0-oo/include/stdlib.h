/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDLIB_H_INCLUDED
#define __ONRAMP_LIBC_STDLIB_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>

#define EXIT_SUCCESS 0
#define EXIT_FAILURE 1

void* malloc(size_t size);
void* calloc(size_t element_count, size_t element_size);
void free(void* p);
void* realloc(void* old_ptr, size_t new_size);

void _Exit(int status);

/**
 * Returns the bounds of the largest region of memory that is currently
 * unallocated.
 *
 * This can be used as the heap for child processes.
 *
 * The memory is not actually allocated. You must not free() this.
 *
 * TODO move this to an internal onramp header
 */
void* __malloc_largest_unused_region(size_t* out_size);

/**
 * Gets the value of the environment variable with the given name.
 */
char* getenv(const char* key);

#endif
