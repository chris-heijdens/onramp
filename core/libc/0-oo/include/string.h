/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STRING_H_INCLUDED
#define __ONRAMP_LIBC_STRING_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>

void* memcpy(void* dest, const void* src, size_t count);
void* memmove(void* dest, const void* src, size_t count);
void* memset(void* dest, int c, size_t count);
int memcmp(const void* a, const void* b, size_t count);

size_t strlen(const char* s);
size_t strnlen(const char* s, size_t max_length);
char* strcpy(char* restrict to, const char* restrict from);
int strcmp(const char* left, const char* right);
char* strchr(const char* s, int c);

// TODO this is only defined in libc/1
char* strrchr(const char* s, int c);

char* strdup(const char* str);
char* strndup(const char* str, size_t max_length);

#endif
