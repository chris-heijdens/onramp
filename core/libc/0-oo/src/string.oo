; The MIT License (MIT)
;
; Copyright (c) 2023-2024 Fraser Heavy Software
;
; Permission is hereby granted, free of charge, to any person obtaining a copy
; of this software and associated documentation files (the "Software"), to deal
; in the Software without restriction, including without limitation the rights
; to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
; copies of the Software, and to permit persons to whom the Software is
; furnished to do so, subject to the following conditions:
;
; The above copyright notice and this permission notice shall be included in all
; copies or substantial portions of the Software.
;
; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
; IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
; FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
; AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
; LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
; SOFTWARE.



; ==========================================================
; size_t strlen(char* str)
; ==========================================================
; Returns the length of the given null-terminated string, not including the
; null-terminator.
;
; This is the standard C strlen() function.
;
; params:
;     - str: r0
; vars:
;     - count: r1
; ==========================================================

=strlen
    ; don't bother to set up a stack frame
    ; we accumulate the count in r1, then move it to r0
    70 81 00 00           ; add r1 0 0           ; zero r1
:__strlen_loop
    7A 8A 80 81           ; ldb ra r0 r1
    7E 8A &__strlen_done  ; jz ra &__strlen_done
    70 81 81 01           ; add r1 r1 1          ; inc r1
    7E 00 &__strlen_loop  ; jz 0 &__strlen_loop
:__strlen_done
    70 80 00 81           ; add r0 0 r1          ; mov r0 r1
    78 8F 00 8C           ; ldw rip 0 rsp        ; ret



; ==========================================================
; char* strcpy(char* dest, const char* src)
; ==========================================================
; Copies null-terminated string src to dest, returning dest.
;
; This is the standard C strcpy() function.
;
; params:
;     - dest: r0
;     - src: r1
; vars:
;     - index: r2
; ==========================================================

=strcpy
    ; don't bother to set up a stack frame
    ; we use an index because it's faster than incrementing both src and dest
    70 82 00 00           ; add r2 0 0           ; zero r2
:__strcpy_loop
    7A 8A 81 82           ; ldb ra r1 r2
    7B 8A 80 82           ; stb ra r0 r2
    7E 8A &__strcpy_done  ; jz ra &__strcpy_done
    70 82 82 01           ; add r2 r2 1          ; inc r2
    7E 00 &__strcpy_loop  ; jz 0 &__strcpy_loop    ; jmp &__strcpy_loop
:__strcpy_done
    ; dest is already in r0
    78 8F 00 8C           ; ldw rip 0 rsp        ; ret



; ==========================================================
; int strcmp(const char* left, const char* right)
; ==========================================================
; Compares null-terminated strings left to right lexicographically. Returns -1
; if left comes before right, 1 if left comes after right, 0 if they are equal.
;
; This is the standard C strcmp() function.
;
; params:
;     - left: r0
;     - right: r1
; vars:
;     - index: r2
; ==========================================================

=strcmp
    ; don't bother to set up a stack frame

    ; we use an index because it's faster than incrementing both src and dest
    70 82 00 00                  ; add r2 0 0           ; zero r2

:__strcmp_loop
    7A 8A 80 82                  ; ldb ra r0 r2
    7A 8B 81 82                  ; ldb rb r1 r2
    7D 8A 8A 8B                  ; cmp ra ra rb
    7E 8A &__strcmp_byte_match   ; jz ra &__strcmp_byte_match
    ; bytes not equal, return the comparison result
    70 80 00 8A                  ; add r0 0 ra          ; mov r0 ra
    78 8F 00 8C                  ; ldw rip 0 rsp        ; ret

:__strcmp_byte_match
    ; bytes match. if zero, strings are equal, return 0
    7E 8B &__strcmp_equal        ; jz rb &__strcmp_equal

    ; otherwise keep looping
    70 82 82 01                  ; add r2 r2 1          ; inc r2
    7E 00 &__strcmp_loop         ; jz 0 &__strcmp_loop    ; jmp &__strcmp_loop

:__strcmp_equal
    ; strings are equal, return 0
    70 80 00 00                  ; add r0 0 0           ; zero r0
    78 8F 00 8C                  ; ldw rip 0 rsp        ; ret



; ==========================================================
; void* memcpy(void* dest, const void* src, size_t count)
; ==========================================================
; Copies count bytes from src to the non-overlapping memory region dest.
;
; This is the standard C memcpy() function.
;
; params:
;     - dest: r0
;     - src: r1
;     - count: r2
; ==========================================================

=memcpy
    ; tail-call memmove
    7C 8A <memmove        ; ims ra <memmove
    7C 8A >memmove        ; ims ra >memmove
    70 8F 8E 8A           ; add rip rpp ra      ; jump



; ==========================================================
; void* memmove(void* dest, const void* src, size_t count)
; ==========================================================
; Copies count bytes from src to the potentially overlapping memory region
; dest.
;
; This is the standard C memmove() function.
;
; params:
;     - dest: r0
;     - src: r1
;     - count: r2
;     - index: r3
; ==========================================================

=memmove
    ; don't bother to set up a stack frame

    ; check whether we should copy down or up
    7D 8A 81 80              ; cmpu ra r1 r0
    7E 8A &__memmove_ret     ; jz ra &__memmove_ret
    70 83 00 00              ; add r3 0 0            ; zero r3
    74 8A 8A FE              ; and ra ra -2
    7E 8A &__memmove_up      ; jz ra &__memmove_up

    ; copy down
:__memmove_down
    7E 82 &__memmove_ret    ; jz r2 &__memmove_ret
    71 82 82 01             ; sub r2 r2 1           ; dec r2
    7A 8A 81 82             ; ldb ra r1 r2
    7B 8A 80 82             ; stb ra r0 r2
    7E 00 &__memmove_down   ; jz 0 &__memmove_down  ; jmp &__memmove_down

    ; copy up
:__memmove_up
    71 8A 82 83             ; sub ra r2 r3
    7E 8A &__memmove_ret    ; jz ra &__memmove_ret
    7A 8A 81 83             ; ldb ra r1 r3
    7B 8A 80 83             ; stb ra r0 r3
    70 83 83 01             ; add r3 r3 1           ; inc r3
    7E 00 &__memmove_up     ; jz 0 &__memmove_up    ; jmp &__memmove_up

    ; done
:__memmove_ret
    ; dest is already in r0
    78 8F 00 8C             ; ldw rip 0 rsp        ; ret



; ==========================================================
; void* memset(void* dest, int c, size_t count)
; ==========================================================
; Sets count bytes in dest to the byte c, returning dest.
;
; This is the standard C memset() function.
;
; params:
;     - dest: r0
;     - src: r1
;     - count: r2
; ==========================================================

=memset
    ; don't bother to set up a stack frame
    7E 82 &__memset_done    ; jz r2 &__memset_done
    71 82 82 01             ; sub r2 r2 1           ; dec r2
    7B 81 80 82             ; stb r1 r0 r2
    7E 00 &memset           ; jz 0 &memset          ; jmp &memset
:__memset_done
    ; dest is already in r0
    78 8F 00 8C             ; ldw rip 0 rsp        ; ret



; ==========================================================
; int memcmp(const char* left, const char* right, size_t count)
; ==========================================================
; Compares left to right regions of count length lexicographically. Returns -1
; if left comes before right, 1 if left comes after right, 0 if they are equal.
;
; This is the standard C memcmp() function.
;
; params:
;     - left: r0
;     - right: r1
;     - count: r2
; vars:
;     - index: r3
; ==========================================================
=memcmp
    70 83 00 00                  ; add r3 0 0           ; zero r3

:__memcmp_loop
    7D 8A 82 83                  ; cmpu ra r2 r3
    7E 8A &__memcmp_equal        ; jz ra &__memcmp_equal
    7A 8A 80 83                  ; ldb ra r0 r3
    7A 8B 81 83                  ; ldb rb r1 r3
    7D 8A 8A 8B                  ; cmpu ra ra rb
    7E 8A &__memcmp_byte_match   ; jz ra &__memcmp_byte_match

    ; bytes not equal, return the comparison result
    70 80 00 8A                  ; add r0 0 ra          ; mov r0 ra
    78 8F 00 8C                  ; ldw rip 0 rsp        ; ret

:__memcmp_byte_match
    ; bytes match, keep looping
    70 83 83 01                  ; add r3 r3 1          ; inc r3
    7E 00 &__memcmp_loop         ; jz 0 &__memcmp_loop    ; jmp &__memcmp_loop

:__memcmp_equal
    ; memory regions are equal, return 0
    70 80 00 00                  ; add r0 0 0           ; zero r0
    78 8F 00 8C                  ; ldw rip 0 rsp        ; ret



; ==========================================================
; char* strchr(const char* s, int c);
; ==========================================================
; Returns a pointer to the first occurrance of c in s, or NULL if s does not
; contain c.
; ==========================================================

=strchr

:strchr_loop
    7A 8A 80 00           ; ldb ra r0 0
    7D 8B 8A 81           ; cmpu rb ra r1
    7E 8B &strchr_match   ; jz rb &strchr_match
    7D 8B 8A 81           ; cmpu rb ra 0
    7E 8B &strchr_match   ; jz rb &strchr_no_match
    70 80 80 01           ; add r0 r0 1             ; inc r0
    7E 00 &strchr_loop    ; jz 0 &strchr_loop       ; jmp &strchr_loop

:strchr_no_match
    70 80 00 00     ; zero r0
    ; fallthrough
:strchr_match
    78 8F 00 8C     ; ldw rip 0 rsp        ; ret
