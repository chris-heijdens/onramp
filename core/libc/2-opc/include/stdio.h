/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2023-2024 Fraser Heavy Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#ifndef __ONRAMP_LIBC_STDIO_H_INCLUDED
#define __ONRAMP_LIBC_STDIO_H_INCLUDED

#ifndef __onramp_libc__
    #error "__onramp/__predef.h must be force-included by the preprocessor before any libc headers."
#endif

#include <__onramp/__size_t.h>
#include <__onramp/__null.h>
#include <__onramp/__va_list.h>

struct __file;
typedef struct __file FILE;
typedef unsigned int fpos_t;

#define _IOFBF 1
#define _IOLBF 2
#define _IONBF 3

#define BUFSIZ 128
#define EOF (-1)
#define FOPEN_MAX 13 // TODO
#define FILENAME_MAX 256
#define L_tmpnam 256 // TODO
#define TMP_MAX 1000 // TODO

#define SEEK_CUR 1
#define SEEK_END 2
#define SEEK_SET 3

extern FILE* __stdin;
extern FILE* __stdout;
extern FILE* __stderr;
#define stdin (__stdin)
#define stdout (__stdout)
#define stderr (__stderr)

// standard C

int remove(const char* filename);
int rename(const char* old, const char* new);
FILE* tmpfile(void);
char* tmpnam(char* s);
int fclose(FILE* stream);
int fflush(FILE* stream);
FILE* fopen(const char* restrict filename, const char* restrict mode);
FILE* freopen(const char* restrict filename, const char* restrict mode, FILE* restrict stream);
int setvbuf(FILE* restrict stream, char* restrict buf, int mode, size_t size);

int fprintf(FILE* restrict stream, const char* restrict format, ...);
int fscanf(FILE* restrict stream, const char* restrict format, ...);
int printf(const char* restrict format, ...);
int scanf(const char* restrict format, ...);
int snprintf(char* restrict buffer, size_t n, const char* restrict format, ...);
int sprintf(char* restrict buffer, const char* restrict format, ...);
int sscanf(const char* restrict buffer, const char* restrict format, ...);
int vfprintf(FILE* restrict stream, const char* restrict format, va_list arg);
int vfscanf(FILE* restrict stream, const char* restrict format, va_list arg);
int vprintf(const char* restrict format, va_list arg);
int vscanf(const char* restrict format, va_list arg);
int vsnprintf(char* restrict buffer, size_t n, const char* restrict format, va_list arg);
int vsprintf(char* restrict buffer, const char* restrict format, va_list arg);
int vsscanf(const char* restrict buffer, const char* restrict format, va_list arg);

//[[deprecated]] char* gets(char* s);
int fgetc(FILE* stream);
char* fgets(char* restrict s, int n, FILE* restrict stream);
int fputc(int c, FILE* stream);
int fputs(const char* restrict s, FILE* restrict stream);
int getc(FILE* stream);
int getchar(void);
int putc(int c, FILE* stream);
int putchar(int c);
int puts(const char* s);
int ungetc(int c, FILE* stream);

size_t fread(void* restrict ptr, size_t size, size_t nmemb, FILE* restrict stream);
size_t fwrite(const void* restrict ptr, size_t size, size_t nmemb, FILE* restrict stream);
int fgetpos(FILE* restrict stream, fpos_t*  restrict pos);
int fseek(FILE* stream, long offset, int whence);
int fsetpos(FILE* stream, const fpos_t* pos);
long ftell(FILE* stream);
void rewind(FILE* stream);

void clearerr(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void perror(const char* s);

// gnu extensions
#ifdef _GNU_SOURCE
int asprintf(char** restrict out_string, const char* restrict format, ...);
int vasprintf(char** restrict out_string, const char* restrict format, va_list args);
#endif

#endif
