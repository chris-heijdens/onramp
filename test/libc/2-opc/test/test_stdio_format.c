// The MIT License (MIT)
// Copyright (c) 2023-2024 Fraser Heavy Software
// This test case is part of the Onramp compiler project.

#define _GNU_SOURCE
#include <stdio.h>

#include <stdlib.h>
#include <string.h>


/*
 * directive parsing
 */

#ifdef DISABLED
static void test_directive_equal(directive_t* actual, directive_t* expected) {
    test_assert(0 == memcmp(actual, expected, sizeof(*actual)));
}

static void test_directive_parse_match(const char* s, directive_t* expected) {
    directive_t actual;
    const char* end = s + strlen(s);
    test_assert(directive_parse(&actual, &s, end));
    test_assert(s == end); // make sure the entire string was consumed
    test_directive_equal(&actual, expected);
}

static void test_stdio_format_directive_parse_s(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.conversion = (uint8_t)'s';
    test_directive_parse_match("%s", &expected);
}

static void test_stdio_format_directive_parse_LS(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.length_modifier = length_modifier_L;
    expected.conversion = (uint8_t)'S';
    test_directive_parse_match("%LS", &expected);
}

static void test_stdio_format_directive_parse_li(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.conversion = (uint8_t)'i';
    expected.length_modifier = length_modifier_l;
    test_directive_parse_match("%li", &expected);
}

static void test_stdio_format_directive_parse_lli(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.conversion = (uint8_t)'i';
    expected.length_modifier = length_modifier_ll;
    test_directive_parse_match("%lli", &expected);
}

static void test_stdio_format_directive_parse_flag_alternate(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.alternate = true;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("%#g", &expected);
}

static void test_stdio_format_directive_parse_flag_zero_padded(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.zero_padded = true;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("%0g", &expected);
}

static void test_stdio_format_directive_parse_flag_left_adjusted(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.left_adjusted = true;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("%-g", &expected);
}

static void test_stdio_format_directive_parse_flag_plus_sign(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.plus_sign = true;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("%+g", &expected);
}

static void test_stdio_format_directive_parse_flag_blank_plus_sign(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.blank_plus_sign = true;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("% g", &expected);
}

static void test_stdio_format_directive_parse_flag_thousands_grouping(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.thousands_grouping = true;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("%'g", &expected);
}

static void test_stdio_format_directive_parse_flag_locale_digits(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.locale_digits = true;
    expected.conversion = (uint8_t)'i';
    test_directive_parse_match("%Ii", &expected);
}

static void test_stdio_format_directive_parse_all_flags(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.alternate = true;
    expected.zero_padded = true;
    expected.left_adjusted = true;
    expected.plus_sign = true;
    expected.blank_plus_sign = true;
    expected.thousands_grouping = true;
    expected.locale_digits = true;
    expected.conversion = (uint8_t)'s';
    test_directive_parse_match("%#0- +'Is", &expected);
}

static void test_stdio_format_directive_parse_position(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.argument_positions = true;
    expected.conversion_argument_position = 4;
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%4$d", &expected);
}

static void test_stdio_format_directive_parse_position_range(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.argument_positions = true;
    expected.conversion_argument_position = 90; // valid digits range from zero to nine
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%90$d", &expected);
}

static void test_stdio_format_directive_parse_field_width(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.field_width = 50;
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%50d", &expected);
}

static void test_stdio_format_directive_parse_precision(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.precision = 70;
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%.70d", &expected);
}

static void test_stdio_format_directive_parse_field_width_by_argument(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.field_width_as_argument = true;
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%*d", &expected);
}

static void test_stdio_format_directive_parse_field_width_by_argument_position(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.argument_positions = true;
    expected.field_width_as_argument = true;
    expected.conversion_argument_position = 5;
    expected.field_width = 14;
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%5$*14$d", &expected);
}

static void test_stdio_format_directive_parse_precision_by_argument(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.precision_as_argument = true;
    expected.conversion = (uint8_t)'f';
    test_directive_parse_match("%.*f", &expected);
}

static void test_stdio_format_directive_parse_precision_by_argument_position(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.argument_positions = true;
    expected.precision_as_argument = true;
    expected.conversion_argument_position = 2;
    expected.precision = 14;
    expected.conversion = (uint8_t)'g';
    test_directive_parse_match("%2$.*14$g", &expected);
}

static void test_stdio_format_directive_parse_field_width_and_precision(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.field_width = 105;
    expected.precision = 99;
    expected.conversion = (uint8_t)'e';
    test_directive_parse_match("%105.99e", &expected);
}

static void test_stdio_format_directive_parse_all_argument_positions(void) {
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.argument_positions = true;
    expected.conversion_argument_position = 101;
    expected.field_width = 2;
    expected.precision = 97;
    expected.field_width_as_argument = true;
    expected.precision_as_argument = true;
    expected.conversion = (uint8_t)'d';
    test_directive_parse_match("%101$*2$.*97$d", &expected);
}

static void test_stdio_format_directive_parse_number_max(void) {
#ifdef DISABLED // TODO test max later
    directive_t expected;
    memset(&expected, 0, sizeof(expected));
    expected.argument_positions = true;
    expected.conversion_argument_position = NUMBER_MAX;
    expected.conversion = (uint8_t)'p';

    // TODO don't use snprintf here!
    char buf[100];
    snprintf(buf, sizeof(buf), "%%%" PRIu64 "$p", (uint64_t)NUMBER_MAX);

    test_directive_parse_match(buf, &expected);
#endif
}
#endif



/*
 * format printing
 */

static void test_stdio_format_printf() {
    printf("Hello\n");
    printf("Hello %s!\n", "world");
    printf("%u\n", 0);
    printf("%u\n", 10);
    printf("%i\n", 10);
    printf("%i\n", -10);
    printf("%d %d %d %d\n", 1, 2, 3, 4);

    printf("%.3d\n", 33);


    // another bug at one point
    char* f;
    if (0 > asprintf(&f, "%s.od", "foo")) {_Exit(1);}
    if (0 != strcmp(f, "foo.od")) {_Exit(1);}
    free(f);
}


int main(void) {

#ifdef DISABLED
    // directive parsing
    test_stdio_format_directive_parse_s();
    test_stdio_format_directive_parse_LS();
    test_stdio_format_directive_parse_li();
    test_stdio_format_directive_parse_lli();
    test_stdio_format_directive_parse_flag_alternate();
    test_stdio_format_directive_parse_flag_zero_padded();
    test_stdio_format_directive_parse_flag_left_adjusted();
    test_stdio_format_directive_parse_flag_plus_sign();
    test_stdio_format_directive_parse_flag_blank_plus_sign();
    test_stdio_format_directive_parse_flag_thousands_grouping();
    test_stdio_format_directive_parse_flag_locale_digits();
    test_stdio_format_directive_parse_all_flags();
    test_stdio_format_directive_parse_position();
    test_stdio_format_directive_parse_position_range();
    test_stdio_format_directive_parse_field_width();
    test_stdio_format_directive_parse_precision();
    test_stdio_format_directive_parse_field_width_by_argument();
    test_stdio_format_directive_parse_field_width_by_argument_position();
    test_stdio_format_directive_parse_precision_by_argument();
    test_stdio_format_directive_parse_precision_by_argument_position();
    test_stdio_format_directive_parse_field_width_and_precision();
    test_stdio_format_directive_parse_all_argument_positions();
    test_stdio_format_directive_parse_number_max();
#endif

    // format printing
    test_stdio_format_printf();
}
