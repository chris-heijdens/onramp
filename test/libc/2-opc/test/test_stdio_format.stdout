Hello
Hello world!
0
10
10
-10
033
